sound/soc/sof/intel/hda.c

```diff
@@ -297,6 +307,268 @@ static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
 	return NULL;
 }
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+static void sof_enable_miscbdcge(struct snd_sof_dev *sdev, bool enable)
+{
+	u32 val;
+
+	val = enable ? PCI_CGCTL_MISCBDCGE_MASK : 0;
+
+	snd_sof_pci_update_bits(sdev, PCI_CGCTL, PCI_CGCTL_MISCBDCGE_MASK, val);
+}
+
+/*
+ * While performing reset, controller may not come back properly causing
+ * issues, so recommendation is to set CGCTL.MISCBDCGE to 0 then do reset
+ * (init chip) and then again set CGCTL.MISCBDCGE to 1
+ */
+static int sof_hda_init_chip(struct snd_sof_dev *sdev, bool full_reset)
+{
+	struct hdac_bus *bus = &sdev->hbus->core;
+	int ret;
+
+	sof_enable_miscbdcge(sdev, false);
+	ret = snd_hdac_bus_init_chip(bus, full_reset);
+	sof_enable_miscbdcge(sdev, true);
+
+	return ret;
+}
+
+irqreturn_t sof_hda_stream_interrupt(int irq, void *context)
+{
+	struct hdac_bus *bus = context;
+	u32 status;
+
+	if (!pm_runtime_active(bus->dev))
+		return IRQ_NONE;
+
+	spin_lock(&bus->reg_lock);
+
+	status = snd_hdac_chip_readl(bus, INTSTS);
+	if (status == 0 || status == 0xffffffff) {
+		spin_unlock(&bus->reg_lock);
+		return IRQ_NONE;
+	}
+
+	/* clear rirb int */
+	status = snd_hdac_chip_readb(bus, RIRBSTS);
+	if (status & RIRB_INT_MASK) {
+		if (status & RIRB_INT_RESPONSE)
+			snd_hdac_bus_update_rirb(bus);
+		snd_hdac_chip_writeb(bus, RIRBSTS, RIRB_INT_MASK);
+	}
+
+	spin_unlock(&bus->reg_lock);
+
+	return snd_hdac_chip_readl(bus, INTSTS) ? IRQ_WAKE_THREAD : IRQ_HANDLED;
+}
+
+/* called from IRQ */
+static void sof_stream_update(struct hdac_bus *bus, struct hdac_stream *hstr)
```

Comment:
that looks wrong. We get the PCM period updates from the existing SOF code, why do we need to do this based on link information?